src/jit/optimizer.cpp

```diff
@@ -4648,7 +4648,8 @@ void Compiler::optCloneLoop(unsigned loopInd, LoopCloneContext* context)
         BasicBlock* newBlk = fgNewBBafter(blk->bbJumpKind, newPred,
                                           /*extendRegion*/ true);
 
-        BasicBlock::CloneBlockState(this, newBlk, blk);
+        bool cloneOk = BasicBlock::CloneBlockState(this, newBlk, blk);
+        noway_assert(cloneOk);
```

Comment:
The cloner's checks for what loops are candidates must be sufficient to guarantee that `gtCloneExpr` will succeed; before this change, its calls to `CloneBlockState` would simply crash in the case that `gtCloneExpr` returned nullptr.  I've added the new return to `CloneBlockState` so that the unroller can handle the false return case, since it uses different candidacy checks (and the existing unroller code already has a bail-out for this exact reason).  I added the noway_assert here to make it explicit that the cloner expects to have made sure the blocks are clonable.

What if I add a comment to that effect ("assert cloneOk because the cloner only considers as candidates expressions that CloneBlockState can handle"), would that address your concern?